js/pathfinding.js

```diff
@@ -5,56 +5,250 @@ import {settingsKey} from "./settings.js";
 import {getSnapPointForTokenObj, getTokenSize, iterPairs} from "./util.js";
 
 import * as GridlessPathfinding from "../wasm/gridless_pathfinding.js";
-import {PriorityQueueSet} from "./data_structures.js";
+import {PriorityQueueSet, ProcessOnceQueue} from "./data_structures.js";
 
+/**
+ * Class to hold all the cached node data, and functions to deal with caching.
+ * 
+ * Since pathfinding can depend on several factors, e.g. the token's size, we keep
+ * several caches, keyed by all the data relevant to pathfinding. If we already have
+ * the maximum number of caches and we need to create another one, we discard the
+ * one not used for the longest.
+ * 
+ * When we select a token, or a token we have selected updates, we start caching
+ * in the background so, when we do start pathfinding, it's very performant.
+ * 
+ * Background caching starts by trying to run an idle process (when the browser is
+ * otherwise not busy), but if it can't do that after an amount of time (e.g. the
+ * CPU is very slow and is busy) then we instead start caching a few nodes each
+ * frame. 
+ */
 class Cache {
 	static maxCacheIds = 5;
+	static maxBackgroundCachingMillis = 10;
+	static maxAnimationCachingMillis = 5;
+	static backgroundCachingTimeoutMillis = 200;
 
 	constructor() {
 		this.nodes = new Map();
 		this.lastUsed = new Map();
+		this.background = {
+			queues: new Map(),
+			nextJobId: null,
+			nextTimeoutId: null,
+			nextAnimationFrameId: null
+		}
 	}
 
 	clear() {
 		this.nodes.clear();
 		this.lastUsed.clear();
+		this.background.queues.clear();
+		if (this.background.nextJobId) {
+			window.cancelIdleCallback(this.background.nextJobId);
+			this.background.nextJobId = null;
+		}
+		this.cancelTimeout();
+		this.cancelAnimationFrame();
 	}
 
 	/**
-	 * Get the cache associated with the given cache ID, creating a new one
-	 * if we don't already have one
+	 * Build a cache ID from the information that could make a difference to the pathfinding algorithm
 	 */
-	getCachedNodes(cacheId) {
+	getCacheId(token) {
+		const cacheData = {};
+
+		// Different-sized tokens snap to different points on the grid,
+		// so they might follow a different path to other tokens
+		cacheData.tokenSize = getTokenSize(token);
+		if (canvas.grid.isHex && game.modules.get("hex-size-support")?.active) {
+			cacheData.hexConfig = {
+				altOrientation: CONFIG.hexSizeSupport.getAltOrientationFlag(token),
+				altSnapping: CONFIG.hexSizeSupport.getAltSnappingFlag(token)
+			}
+		}
+
+		// If levels is enabled, the token's elevation can affect which walls
+		// they need to worry about
+		if (game.modules.get("levels")?.active) {
+			cacheData.elevation = token.data.elevation;
+		}
+
+		const cacheId = JSON.stringify(cacheData);
+		this.ensureCacheExists(cacheId);
```

Comment:
`ensureCacheExists` sounds a bit out of place for a function that calculates a cache id. Maybe we can find a better spot?